### DISCUSSION TITLE
alpha values variable with interval colormaps

### DISCUSSION THREAD

**Discussion by gagecarto** [developmentseed/titiler#525] [Q&A] [answered]
I am working to filter a raster between variable range intervals as chosen by the user. To start I've been doing this using colormaps like below where I just make the upper and lower ranges transparent
`myColormap=[
   [[0,0.21],[0,0,0,0]],
   [[0.21,0.75],[22,37,194,255]],
   [[0.75,1],[0,0,0,0]]
]`

This works great and tiles are returned as transparent in those areas where values are within the lower and upper bounds. E.g. below
![image](https://user-images.githubusercontent.com/5776766/197644023-ea3a116d-7953-4056-8c71-e5bd54c1ad15.png)
`https://titiler.xyz/cog/tiles/WebMercatorQuad/10/200/383@1x?url=myfilelocation&bidx=1&colormap=[[[0,0],[0,0,0,0]],[[0,0.17],[12,208,32,255]],[[0.17,1],[0,0,0,0]]]`

For some reason there is variability in the tiles that are returned and sometimes tiles are not transparent where they should be as defined in the color map. The call below returns a tile with no opacity but just black in those areas where the alpha value should be 0. 
![image](https://user-images.githubusercontent.com/5776766/197644272-dc9a79d2-36a2-4db2-81a3-7e09546eac2c.png)
`https://titiler.xyz/cog/tiles/WebMercatorQuad/10/202/377@1x?url=myfilelocation&bidx=1&maxzoom=14&colormap=[[[0,0.29],[0,0,0,0]],[[0.25,0.75],[22,37,194,255]],[[0.75,1],[0,0,0,0]]]`

Any ideas why zoom would effect the alpha value on those tile requests?

**Comment by gagecarto** [comment, marked as answer]
I got this sorted out. I had my tile format set to automatic and titiler would sometimes send jpg. Adding _&format=png_ cleared things up

`https://titiler.xyz/cog/tiles/WebMercatorQuad/10/202/377@1x?url=myfilelocation&bidx=1&maxzoom=14&colormap=[[[0,0.29],[0,0,0,0]],[[0.25,0.75],[22,37,194,255]],[[0.75,1],[0,0,0,0]]]&format=png`

**Reply by vincentsarago** [reply]
Oh that's an edge case were we create a new `mask` from the colormap :-(

**Reply by gagecarto** [reply]
@vincentsarago - I am actually still seeing some tiles returned which don't follow the alpha values. 

Are you saying that this is an issue associated with edges and new masks? Any suggestions for forcing the tiles to always have the alpha values?

**Reply by vincentsarago** [reply]
I'm not sure but this is mostly where it come from https://github.com/cogeotiff/rio-tiler/blob/main/rio_tiler/utils.py#L444-L449

would you be able to share the file? ( you can use https://github.com/cogeotiff/rio-faux to impersonate the file)

**Reply by gagecarto** [reply]
See the faux file below. 
https://www.dropbox.com/s/k0c3r1wbdtm26yx/test.tif?dl=0

**Reply by vincentsarago** [reply]
@gagecarto well I can't really reproduce this. When using discrete colormap you are creating a new Alpha band which then should be taken into account when creating the PNG. 

Sadly I the `faux` file is only nodata (bug/missing feature in rio-faux 🤦) so I tried with another file:

<img width="944" alt="Screen Shot 2022-10-26 at 9 24 54 AM" src="https://user-images.githubusercontent.com/10407788/197961385-9c6f7d30-e9fc-45cd-8ea6-1ca729cdf7a0.png">

**Reply by gagecarto** [reply]
@vincentsarago - It's all good today. I think some of those other tiles were still cached. Forcing to png fixed it

**Reply by vincentsarago** [reply]
🥳
